### PR TITLE
fabtests/examples: Add libfabric examples

### PR DESCRIPTION
**Pull request by Juee14Desai**
Add examples to explain basic libfabric apis. Four examples are added:

example_rdm.c:
Explain simple message exchange between server and client using RDM endpoint.

example_rdm_oob.c:
This example explains simple meesage exchange with out of band addressing capability.

example_rdm_rma.c:
Explain message exchange using remote memory access. Two operations are explained
here, remote write and remote read. In local write operation, client is going to write data
from its local write buffer to the servers write buffer. In remote read operation, the server
is going to read data from client’s read buffer into its own local read buffer.

example_rdm_tagged.c:
This example demonstrates how tagged messages work and specifically how the messages
are received in the same order as they were sent.

### CODE REVIEW

File: fabtests/examples/example_rdm.c [hunk 1]
```diff
@@ -0,0 +1,507 @@
...
+	 */
+	hints->addr_format = FI_SOCKADDR_IN;
+
+	/*
+	 * Default to FI_DELIVERY_COMPLETE which will make sure completions do
+	 * not get generated until our message arrives at the destination.
+	 * Otherwise, the client might get a completion and exit before the
+	 * server receives the message. This is to make the test simpler.
+	 */
+	hints->tx_attr->op_flags = FI_DELIVERY_COMPLETE;
```
Comment (shefty) [inline fabtests/examples/example_rdm.c:495]: I would move the initialization of hints into a function that is placed above 'initialization'.  This is the first step that an app needs to take.

File: fabtests/examples/example_rdm.c [hunk 2]
```diff
@@ -0,0 +1,507 @@
...
+static int initialize(void)
+{
+	struct fi_cq_attr cq_attr = {0};
+	struct fi_av_attr av_attr = {0};
+	const struct sockaddr_in *sin;
+	char str_addr[INET_ADDRSTRLEN];
+	int ret;
+
+	/*
+	 * The first OFI call to happen for initialization is fi_getinfo which
```
Comment (shefty) [inline fabtests/examples/example_rdm.c:68]: The first call is usually fi_allocinfo() for the hints.  See comment below about moving hints initialization above this function.

File: fabtests/examples/example_rdm.c [hunk 3]
```diff
@@ -0,0 +1,507 @@
...
+	 * address/port refer to source information. This is not set for the
+	 * client because the fields refer to the server, not the caller
+	 * (client).
+	 */
+
+	/*
+	 * (fi_getinfo)(uint32_t version, const char *node,
+	 * 		const char *service, uint64_t flags,
+	 * 		const struct fi_info *hints, struct fi_info **info)
+	 */
```
Comment (shefty) [inline fabtests/examples/example_rdm.c:84]: I'd remove all comments that simply list the call like this.  The call itself should be sufficient.

File: fabtests/examples/example_rdm.c [hunk 4]
```diff
@@ -0,0 +1,507 @@
...
+	 * hardware and software resources that access a single physical or
+	 * virtual network. All network ports on a system that can communicate
+	 * with each other through their attached networks belong to the same
+	 * fabric.
+	 */
+
+	/*
+	 * fi_fabric(struct fi_fabric_attr *attr,
+	 * 	     struct fid_fabric **fabric, void *context);
+	 */
```
Comment (shefty) [inline fabtests/examples/example_rdm.c:120]: drop comment

File: fabtests/examples/example_rdm.c [hunk 5]
```diff
@@ -0,0 +1,507 @@
...
+
+	/*
+	 * Initialize our domain (associated with our fabric). A domain defines
+	 * the boundary for associating different resources together.
+	 */
+
+	/*
+	 * fi_domain(struct fid_fabric *fabric, struct fi_info *info,
+	 * 	     struct fid_domain **domain, void *context);
+	 */
```
Comment (shefty) [inline fabtests/examples/example_rdm.c:135]: drop comment

File: fabtests/examples/example_rdm.c [hunk 6]
```diff
@@ -0,0 +1,507 @@
...
+	 * are three main types of endpoints:
+	 * FI_EP_MSG - connected, reliable
+	 * FI_EP_RDM - unconnected, reliable
+	 * FI_EP_DGRAM - unconnected, unreliable
+	 * The type of endpoint will be requested in hints/fi_getinfo.
+	 * Different providers support different types of endpoints. TCP
+	 * supports only FI_EP_MSG but when used with RxM, can support
+	 * FI_EP_RDM. In this application, we requested TCP and FI_EP_RDM
+	 * so we will automatically pick up RxM in order to use RDM with TCP
+	 * hardware.
```
Comment (shefty) [inline fabtests/examples/example_rdm.c:154]: Stop comment at Different providers support different types of endpoints.  The rest of the comment is no longer valid.

File: fabtests/examples/example_rdm.c [hunk 7]
```diff
@@ -0,0 +1,507 @@
...
+	 */
+
+	/*
+	 * fi_endpoint(struct fid_domain *domain, struct fi_info *info,
+	 * 	       struct fid_ep **ep, void *context);
+	 */
```
Comment (shefty) [inline fabtests/examples/example_rdm.c:160]: drop comment

File: fabtests/examples/example_rdm.c [hunk 8]
```diff
@@ -0,0 +1,507 @@
...
+	 * Initialize our completion queue. Completion queues are used to
+	 * report events associated with data transfers. In this example, we
+	 * use one CQ that tracks sends and receives, but often times there
+	 * will be separate CQs for sends and receives.
+	 */
+
+	/*
+	 * fi_cq_open(struct fid_domain *domain, struct fi_cq_attr *attr,
+	 * 	      struct fid_cq **cq, void *context);
+	 */
```
Comment (shefty) [inline fabtests/examples/example_rdm.c:177]: drop comment

File: fabtests/examples/example_rdm.c [hunk 9]
```diff
@@ -0,0 +1,507 @@
...
+	if (ret) {
+		printf("fi_ep_bind cq error (%d)\n", ret);
+		return ret;
+	}
+
+	/*
+	 * Initialize our address vector. Address vectors are used to map
+	 * higher level addresses, which may be more natural for an application
+	 * to use, into fabric specific addresses. An AV_TABLE av will map
+	 * thse addresses to indexed addresses, starting with fi_addr 0. These
```
Comment (shefty) [inline fabtests/examples/example_rdm.c:204]: thse -> these

File: fabtests/examples/example_rdm.c [hunk 10]
```diff
@@ -0,0 +1,507 @@
...
+	 * addresses are used in data transfer calls to specify which peer to
+	 * send to/recv from. Address vectors are only used for FI_EP_RDM and
+	 * FI_EP_DGRAM endpoints, allowing the application to avoid connection
+	 * management. For FI_EP_MSG endpoints, the AV is replaced by the
+	 * traditional TCP listen/connect/accept steps.
```
Comment (shefty) [inline fabtests/examples/example_rdm.c:209]: remove 'TCP' from comment

File: fabtests/examples/example_rdm.c [hunk 11]
```diff
@@ -0,0 +1,507 @@
...
+	ret = fi_enable(ep);
+	if (ret) {
+		printf("fi_enable error (%d)\n", ret);
+		return ret;
+	}
+
+	return 0;
+}
+
+/* All OFI resources are cleaned up using the same fi_close(fid) call. */
```
Comment (shefty) [inline fabtests/examples/example_rdm.c:273]: Expand comment to indicate that resources must be closed in a specific order to allow references between objects to be removed correctly.  E.g. the EP must be closed before the CQ or AV.

File: fabtests/examples/example_rdm.c [hunk 12]
```diff
@@ -0,0 +1,507 @@
...
+static void cleanup(void)
+{
+	int ret;
+
+	/* fi_close(struct fid *ep); */
```
Comment (shefty) [inline fabtests/examples/example_rdm.c:278]: remove comments like these -- they end up being more distracting than helpful.

File: fabtests/examples/example_rdm.c [hunk 13]
```diff
@@ -0,0 +1,507 @@
...
+struct fid_fabric *fabric = NULL;
+struct fid_domain *domain = NULL;
+struct fid_ep *ep = NULL;
+struct fid_av *av = NULL;
+struct fid_cq *cq = NULL;
+char buf[BUF_SIZE];
+fi_addr_t fi_addr = FI_ADDR_UNSPEC;
+
+/*
+ * Initializes all basic OFI resources to allow for a server/client to
```
Comment (shefty) [inline fabtests/examples/example_rdm.c:56]: Replace 'OFI' with 'libfabric everywhere.

File: fabtests/examples/example_rdm.c [hunk 14]
```diff
@@ -0,0 +1,507 @@
...
+char *dst_addr = NULL;
+char *port = "9228";
+struct fi_info *hints, *info;
+struct fid_fabric *fabric = NULL;
+struct fid_domain *domain = NULL;
+struct fid_ep *ep = NULL;
+struct fid_av *av = NULL;
+struct fid_cq *cq = NULL;
+char buf[BUF_SIZE];
+fi_addr_t fi_addr = FI_ADDR_UNSPEC;
```
Comment (shefty) [inline fabtests/examples/example_rdm.c:53]: The above can be declared as static.

File: fabtests/examples/example_rdm.c [hunk 15]
```diff
@@ -0,0 +1,507 @@
...
+ * a send needs a valid fi_addr as input to tell the provider where to send the
+ * message. Similar to the receive buffer posting porcess, when posting a send
+ * buffer, if the provider is not ready to process messages, it may return
+ * -FI_EAGAIN. This does not indicate an error, but rather that the application
+ * should try again later. Just like the receive, we drive progress with
+ * fi_cq_read if this is the case.
+ */
+static int post_send(void)
+{
+	char *msg = "Hello, server! I am the client you've been waiting for!\0";
```
Comment (shefty) [inline fabtests/examples/example_rdm.c:358]: \0 at end of string is not needed.

File: fabtests/examples/example_rdm.c [hunk 16]
```diff
@@ -0,0 +1,507 @@
...
+	int ret;
+
+	(void) snprintf(buf, BUF_SIZE, "%s", msg);
```
Comment (shefty) [inline fabtests/examples/example_rdm.c:361]: There's not really a reason to copy a string to another buffer just to send the data.  Just pass msg into fi_send().

File: fabtests/examples/example_rdm.c [hunk 17]
```diff
@@ -0,0 +1,507 @@
...
+
+	do {
+		/*
+		 * fi_send(struct fid_ep *ep, const void *buf, size_t len,
+		 *	   void *desc, fi_addr_t dest_addr,
+		 *	   void *context);
+		 */
+		ret = fi_send(ep, buf, BUF_SIZE, NULL, fi_addr, NULL);
```
Comment (shefty) [inline fabtests/examples/example_rdm.c:369]: Note that this assumes an inject size >= sizeof(msg), or that memory registration isn't needed.  That should be captured in the hints (if assuming inject_size), or a comment (if we're not supporting memory registration).

File: fabtests/examples/example_rdm.c [hunk 18]
```diff
@@ -0,0 +1,507 @@
...
+	} while (ret);
+
+	return 0;
+}
+
+/*
+ * Post a send buffer. This call does not ensure a message has been sent, just
+ * that a buffer has been submitted to OFI to be sent. Unlike a receive buffer,
+ * a send needs a valid fi_addr as input to tell the provider where to send the
+ * message. Similar to the receive buffer posting porcess, when posting a send
```
Comment (shefty) [inline fabtests/examples/example_rdm.c:350]: porcess -> process

File: fabtests/examples/example_rdm.c [hunk 19]
```diff
@@ -0,0 +1,507 @@
...
+/*
+ * Wait for the message to be sent/received using the CQ. fi_cq_read not only
+ * drives progress but also returns any completed events to notify the
+ * application that it can reuse the send/recv buffer. The returned completion
+ * entry will have fields set to let the application know what operation
+ * completed. Not all fields will be valid. The fields set will be indicated by
+ * the cq format (when creating the CQ). In this example, we use
+ * FI_CQ_FORMAT_MSG in order to use the flags field.
+ */
+static int wait_cq(void)
```
Comment (shefty) [inline fabtests/examples/example_rdm.c:390]: rename to spin_for_comp()

**Comment by Juee14Desai** [comment]
Hi @shefty,

Please review these examples. I am quite late for the PR but I think these are in much better shape than when I presented. I am not able to request review from you so tagging you here.

Thanks,
Juee

File: fabtests/examples/example_rdm.c [hunk 20]
```diff
@@ -0,0 +1,507 @@
...
+	if (!hints)
+		return EXIT_FAILURE;
+
+	/*
+	 * Server run with no args, client has server's address as an
+	 * argument.
+	 */
+	dst_addr = argv[optind];
+
+//Set anything in hints that the application needs
```
Comment (shefty) [inline fabtests/examples/example_rdm.c:465]: outdented comments should be removed or indented.  There's also a mix of comment format.

File: fabtests/examples/example_rdm.c [hunk 21]
```diff
@@ -0,0 +1,507 @@
...
+	 * listen/connect/accept.
+	 */
+	hints->ep_attr->type = FI_EP_RDM;
+
+	/*
+	 * Request basic messaging capabilities from the provider (no tag
+	 * matching, no RMA, no atomic operations)
+	 */
+	hints->caps = FI_MSG;
+
```
Comment (shefty) [inline fabtests/examples/example_rdm.c:479]: I would add setting hints->mode = 0 and hints->domain_attr->mr_mode = 0, with comments explaining what the modes do

**Review by shefty — CHANGES_REQUESTED** [review]
See comments on example_rdm.c only.  Other tests were not reviewed.

**Comment by Juee14Desai** [comment]
Only changed example_rdm.c, others are pending.

**Comment by Juee14Desai** [comment]
All the examples are updated according to feedback received on example_rdm.c.

**Comment by zachdworkin** [comment]
@Juee14Desai You need to rebase to pickup the cloudbees CI files

**Pull request closed without merging**